src/java.base/share/classes/java/io/SerializationMisdeclarationChecker.java

```diff
@@ -0,0 +1,300 @@
+            Class<?>[] paramTypes, Class<?> retType) {
+        for (Class<?> cls = cl; cls != null; cls = cls.getSuperclass()) {
+            for (Method m : cls.getDeclaredMethods()) {
+                if (m.getName().equals(name)) {
+                    privilegedCheckAccessibleMethod(cls, m, paramTypes, retType);
+                }
+            }
+        }
+    }
+
+    private void privilegedCheckAccessibleMethod(Class<?> cls, Method m,
+            Class<?>[] paramTypes, Class<?> retType) {
+        if (cls.isEnum()) {
+            commitEvent(cls, ACC_METH_INEFFECTIVE_ENUM,
+                    m + " is not effective on an enum class");
+        }
+        if (isAbstract(m)) {
+            commitEvent(ACC_METH_NON_ABSTRACT,
+                    m + " must be non-abstract to be effective");
+        }
+        if (isStatic(m)) {
+            commitEvent(ACC_METH_NON_STATIC,
+                    m + " must be non-static to be effective");
+        }
+        if (m.getReturnType() != retType) {
+            commitEvent(ACC_METH_RET_TYPE,
+                    m + " must have return type " + retType + " to be effective");
+        }
+        if (!Arrays.equals(m.getParameterTypes(), paramTypes)) {
+            commitEvent(ACC_METH_PARAM_TYPES,
+                    m + " must have parameter types " + Arrays.toString(paramTypes) + " to be effective");
+        }
+        if (isPrivate(m) && cl != cls
+                || isPackageProtected(m) && !isSamePackage(cl, cls)) {
+            commitEvent(ACC_METH_NON_ACCESSIBLE,
+                    m + " is not accessible");
+        }
+    }
+
+    private static boolean isSamePackage(Class<?> cl0, Class<?> cl1) {
+        return cl0.getClassLoader() == cl1.getClassLoader()
+                && cl0.getPackageName().equals(cl1.getPackageName());
+    }
+
+    private boolean isOrdinaryClass() {
+        /* class Enum and class Record are not considered ordinary classes */
+        return !(cl.isRecord() || cl.isEnum() || cl.isArray()
+                || Enum.class == cl || Record.class == cl
+                || Proxy.isProxyClass(cl));
+    }
+
+    private static boolean isPrivate(Member m) {
+        return (m.getModifiers() & PRIVATE) != 0;
+    }
+
+    private static boolean isPackageProtected(Member m) {
+        return (m.getModifiers() & (PRIVATE | PROTECTED | PUBLIC)) == 0;
+    }
+
+    private static boolean isAbstract(Member m) {
+        return (m.getModifiers() & ABSTRACT) != 0;
+    }
+
+    private static boolean isFinal(Member m) {
+        return (m.getModifiers() & FINAL) != 0;
+    }
+
+    private static boolean isStatic(Member m) {
+        return (m.getModifiers() & STATIC) != 0;
+    }
+
+    private static Field declaredField(Class<?> cl, String name) {
+        try {
+            return cl.getDeclaredField(name);
+        } catch (NoSuchFieldException ignored) {
+        }
+        return null;
+    }
+
+    private static Object getObject(Field f) {
```

Comment:
Should we call this method `getStaticFieldValue(...)`, because that's what the implementation is.

Reply:
Since the only parameter is `Field`, it has to be a static field almost for sure.
Further, there's a `getLong()` method down below that operates on a static field as well and that one would also need a renaming.
If I can come up with better names they will be in the next commit.